meteor/server/api/deviceTriggers/PieceInstancesObserver.ts

```diff
@@ -0,0 +1,99 @@
+import { Meteor } from 'meteor/meteor'
+import { RundownPlaylistActivationId, ShowStyleBaseId } from '@sofie-automation/corelib/dist/dataModel/Ids'
+import { RundownPlaylists, ShowStyleBases, PieceInstances, PartInstances } from '../../collections'
+import { logger } from '../../logging'
+import { rundownPlaylistFieldSpecifier } from './reactiveContentCache'
+import {
+	ContentCache,
+	createReactiveContentCache,
+	partInstanceFieldSpecifier,
+	pieceInstanceFieldSpecifier,
+} from './reactiveContentCacheForPieceInstances'
+import { waitForAllObserversReady } from '../../publications/lib/lib'
+
+const REACTIVITY_DEBOUNCE = 20
+
+type ChangedHandler = (cache: ContentCache) => () => void
+
+export class PieceInstancesObserver {
+	#observers: Meteor.LiveQueryHandle[] = []
+	#cache: ContentCache
+	#cancelCache: () => void
+	#cleanup: (() => void) | undefined
+	#disposed = false
+
+	constructor(onChanged: ChangedHandler) {
+		const { cache, cancel: cancelCache } = createReactiveContentCache(() => {
+			this.#cleanup = onChanged(cache)
+			if (this.#disposed) this.#cleanup()
+		}, REACTIVITY_DEBOUNCE)
+
+		this.#cache = cache
+		this.#cancelCache = cancelCache
+	}
+
+	static async create(
+		activationId: RundownPlaylistActivationId,
+		showStyleBaseId: ShowStyleBaseId,
+		onChanged: ChangedHandler
+	): Promise<PieceInstancesObserver> {
+		logger.silly(`Creating PieceInstancesObserver for activationId "${activationId}"`)
+
+		const observer = new PieceInstancesObserver(onChanged)
+
+		await observer.initObservers(activationId, showStyleBaseId)
+
+		return observer
+	}
+
+	private async initObservers(activationId: RundownPlaylistActivationId, showStyleBaseId: ShowStyleBaseId) {
+		this.#observers = await waitForAllObserversReady([
+			RundownPlaylists.observeChanges(
+				{
+					activationId,
+				},
+				this.#cache.RundownPlaylists.link(),
+				{
+					projection: rundownPlaylistFieldSpecifier,
+				}
+			),
+			ShowStyleBases.observeChanges(showStyleBaseId, this.#cache.ShowStyleBases.link()),
+			PieceInstances.observeChanges(
+				{
+					playlistActivationId: activationId,
+					reset: { $ne: true },
+					disabled: { $ne: true },
+					reportedStoppedPlayback: { $exists: false },
+					'piece.virtual': { $ne: true },
+				},
+				this.#cache.PieceInstances.link(),
+				{
+					projection: pieceInstanceFieldSpecifier,
+				}
+			),
+			PartInstances.observeChanges(
+				{
+					playlistActivationId: activationId,
+					reset: { $ne: true },
+					'timings.reportedStoppedPlayback': { $ne: true },
+				},
+				this.#cache.PartInstances.link(),
+				{
+					projection: partInstanceFieldSpecifier,
+				}
+			),
+		])
+	}
+
+	public get cache(): ContentCache {
+		return this.#cache
+	}
+
+	public stop = (): void => {
+		this.#disposed = true
+		this.#cancelCache()
+		this.#observers.forEach((observer) => observer.stop())
+		this.#cleanup?.()
+		this.#cleanup = undefined
```

Comment:
Just a nitpick. If you assign this to undefined, the mental logic of this would be that the value was never set. However, if you assign it to null (a falsy value that works similar to undefined as you know) it implies more that the function resets the parameter.

Reply:
> However, if you assign it to null (a falsy value that works similar to undefined as you know)

True, but I `undefined` meaning 'unset' is an established pattern in this codebase, and it doesn't really matter if the function resets the parameter or not - the behavior should be the same regardless if this is the value hasn't been set before or if it has been reset. With that said, since `typeof null` is `object`, using `undefined` gives less chance of it being mistaken for something else, even if it's falsy.